File: logi-plugin-sdk-python/library.py
from logi_plugin_pb2 import Envelope
from logi_plugin_pb2 import ManagerHello
from logi_plugin_pb2 import PluginActionConfigurationSchemeResponse
from logi_plugin_pb2 import PluginHello
from logi_plugin_pb2 import Pong
from logi_plugin_pb2 import ResponseInfo
from logi_plugin_pb2 import RetryAfter
from logi_plugin_pb2 import ReleaseAction
from logi_plugin_pb2 import VisibilityChanged
from logi_plugin_pb2 import UpdateAnalogControl
from logi_plugin_pb2 import Settings
from logi_plugin_pb2 import LogEvent
from logi_plugin_pb2 import SetData
from logi_plugin_pb2 import GetData
from logi_plugin_pb2 import WipeData
from logi_plugin_pb2 import SavedData
from logi_plugin_pb2 import OauthResponse
from logi_plugin_pb2 import SettingsRequest
from logi_plugin_pb2 import GetAnalogControlValue
from logi_plugin_pb2 import AnalogControlStatus
from logi_plugin_pb2 import ActionCellStatus
from logi_plugin_pb2 import UserInfo
from logi_plugin_pb2 import InitLogin
from logi_plugin_pb2 import Logout
from logi_plugin_pb2 import AnalogControlStatusList
from logi_plugin_pb2 import VisibilityChangedList
from logi_plugin_pb2 import GetUserInfoList
from logi_plugin_pb2 import GetActionStatus
from logi_plugin_pb2 import ActionStatusList
from logi_plugin_pb2 import ActionStatus
import json

class Library:
    def __init__(self):
        self.mapper = {
        'ManagerHello': self.onHello,
            'Ping': self.onPing,
            'TriggerAction': self.onTriggerAction,
            'PluginActionConfigurationSchemeRequest': self.onPluginActionConfigurationSchemeRequest,
            'PluginActionConfigurationResponse': self.onPluginActionConfigurationResponse,
            'RetryAfter': self.onRetryAfter,
            'ReleaseAction': self.onReleaseAction,
            'VisibilityChanged': self.onVisibilityChanged,
            'UpdateAnalogControl': self.on_updateAnalogControl,
            'Settings': self.onSettings,
            'SavedData': self.onSavedData,
            'OauthResponse': self.onOauthResponse,
            'GetAnalogControlValue': self.onGetAnalogControlValue,
            'InitLogin': self.on_initLogin,
            'Logout': self.oLogout,
            'ResponseInfo': self.onResponseInfo,
            'AnalogControlStatus': self.onAnalogControlStatus,
            'AnalogControlStatusList': self.onAnalogControlStatusList,
            'VisibilityChangedList': self.onVisibilityChangedList,
            'ActionCellStatus': self.onActionCellStatus,
            'UserInfo': self.onUserInfo,
            'PluginActionConfigurationSchemeResponse': self.onPluginActionConfigurationSchemeResponse,
            'GetUserInfoList': self.onGetUserInfoList,
            'GetActionStatus': self.onGetActionStatus,
            'ActionStatusList': self.onActionStatusList,
            'ActionStatus': self.onActionStatus
        }

    __getProto = lambda type, obj: (
    {
        'Envelope': lambda obj: Envelope(obj),
        'PluginHello': lambda obj: PluginHello(obj),
        'Pong': lambda obj: Pong(obj),
        'Settings': lambda obj: Settings(obj),
        'LogEvent': lambda obj: LogEvent(obj),
        'SetData': lambda obj: SetData(obj),
        'GetData': lambda obj: GetData(obj),
        'SettingsRequest': lambda obj: SettingsRequest(obj),
        'WipeData': lambda obj: WipeData(obj),
        'ResponseInfo': lambda obj: ResponseInfo(obj),
        'PluginActionConfigurationSchemeResponse': lambda obj: PluginActionConfigurationSchemeResponse(obj),
        'AnalogControlStatus': lambda obj: AnalogControlStatus(obj),
        'ActionCellStatus': lambda obj: ActionCellStatus(obj),
        'UserInfo': lambda obj: UserInfo(obj),
        'AnalogControlStatusList': lambda obj: AnalogControlStatusList(obj),
        'VisibilityChangedList': lambda obj: VisibilityChangedList(obj),
        'GetActionStatus': lambda obj: GetActionStatus(obj),
        'ActionStatusList': lambda obj: ActionStatusList(obj),
        'ActionStatus': lambda obj: ActionStatus(obj)
    }.get(type, lambda obj: None)(obj))

    getType = lambda obj, typeName: (
        {
            **obj,
            "@type": f"type.googleapis.com/logi.plugin.protocol.{typeName}"
        })

    __getProtoMessage = lambda self, protoName, messageId, obj={}: (
    protoMessage := self.getProto(protoName, obj),
    envelop := self.getProto('Envelope', {'id': messageId}),
    setattr(envelop, 'message', protoMessage),
    envelopObj := json.loads(envelop.toJsonString()),
    setattr(envelopObj, 'message', self.getType(protoMessage, protoName)),
    json.dumps(envelopObj))

    __getProtoMessageWithResponse = lambda self, protoName, messageId, obj={}, err={}: (
        envelop := self.getProto('Envelope', {'id': messageId}),
        envelopObj := json.loads(envelop.toJsonString()),
        setattr(envelopObj, 'message', self.getType(self.getProto(protoName, obj), protoName)) if protoName != 'ResponseInfo' else None,
        setattr(envelopObj, 'response', {'code': err['code'] if 'code' in err else 1, 'what': err['what'] if 'what' in err else ''}),
        json.dumps(envelopObj))

    def __defaultResponse(self, msg, ws):
        response = self.__getProtoMessageWithResponse('ResponseInfo', msg['id'], {}, {
            'code': -5,
            'what': 'Not Implemented'
        })
        try:
            ws.send(response)
        except Exception as e:
            print('Something went wrong with ResponseInfo', e)

    def onHello(self, msg):
        return

    def __onPing(self, msg, ws):
        response = self.__getProtoMessageWithResponse('Pong', msg['id'])
        try:
            ws.send(response)
        except Exception as e:
            print('Something went wrong with Pong', e)

    def onTriggerAction(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onPluginActionConfigurationSchemeRequest(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onPluginActionConfigurationResponse(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onRetryAfter(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onReleaseAction(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onVisibilityChanged(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onUpdateAnalogControl(self, msg, ws):
        self.defaultResponse(msg, ws)

    def onSettings(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onSavedData(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onOauthResponse(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onGetAnalogControlValue(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onInitLogin(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def on_logout(self, msg, ws):
        self.default_response(msg, ws)

    def onResponseInfo(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onAnalogControlStatus(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onAnalogControlStatusList(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onVisibilityChangedList(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onActionCellStatus(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onUserInfo(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onPluginActionConfigurationSchemeResponse(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onGetUserInfoList(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onGetActionStatus(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onActionStatusList(self, msg, ws):
        self.__defaultResponse(msg, ws)

    def onActionStatus(self, msg, ws):
        self.__defaultResponse(msg, ws)

File: sdk/library.js
import {
  Envelope,
  ManagerHello,
  PluginActionConfigurationSchemeResponse,
  PluginHello,
  Pong,
  ResponseInfo,
  RetryAfter,
  ReleaseAction,
  VisibilityChanged,
  UpdateAnalogControl,
  Settings,
  LogEvent,
  SetData,
  GetData,
  WipeData,
  SavedData,
  OauthResponse,
  SettingsRequest,
  GetAnalogControlValue,
  AnalogControlStatus,
  ActionCellStatus,
  UserInfo,
  InitLogin,
  Logout,
  AnalogControlStatusList,
  VisibilityChangedList,
  GetUserInfoList,
  GetActionStatus,
  ActionStatusList,
  ActionStatus
} from './schema/logi_plugin_pb.js';

class Library {
  constructor() {
    this.mapper = {
      ['ManagerHello']: (msg) => {
        this.onHello(msg);
      },
      ['Ping']: (msg, ws) => {
        this.#onPing(msg, ws);
      },
      ['TriggerAction']: (msg, ws) => {
        this.onTriggerAction(msg, ws);
      },
      ['PluginActionConfigurationSchemeRequest']: (msg, ws) => {
        this.onPluginActionConfigurationSchemeRequest(msg, ws);
      },
      ['PluginActionConfigurationResponse']: (msg, ws) => {
        this.onPluginActionConfigurationResponse(msg, ws);
      },
      ['RetryAfter']: (msg, ws) => {
        this.onRetryAfter(msg, ws);
      },
      ['ReleaseAction']: (msg, ws) => {
        this.onReleaseAction(msg, ws);
      },
      ['VisibilityChanged']: (msg, ws) => {
        this.onVisibilityChanged(msg, ws);
      },
      ['UpdateAnalogControl']: (msg, ws) => {
        this.onUpdateAnalogControl(msg, ws);
      },
      ['Settings']: (msg, ws) => {
        this.onSettings(msg, ws);
      },
      ['SavedData']: (msg, ws) => {
        this.onSavedData(msg, ws);
      },
      ['OauthResponse']: (msg, ws) => {
        this.onOauthResponse(msg, ws);
      },
      ['GetAnalogControlValue']: (msg, ws) => {
        this.onGetAnalogControlValue(msg, ws);
      },
      ['InitLogin']: (msg, ws) => {
        this.onInitLogin(msg, ws);
      },
      ['ResponseInfo']: (msg, ws) => {
        this.onResponseInfo(msg, ws);
      },
      ['Logout']: (msg, ws) => {
        this.onLogout(msg, ws);
      },
      ['AnalogControlStatus']: (msg, ws) => {
        this.onAnalogControlStatus(msg, ws);
      },
      ['AnalogControlStatusList']: (msg, ws) => {
        this.onAnalogControlStatusList(msg, ws);
      },
      ['VisibilityChangedList']: (msg, ws) => {
        this.onVisibilityChangedList(msg, ws);
      },
      ['ActionCellStatus']: (msg, ws) => {
        this.onActionCellStatus(msg, ws);
      },
      ['UserInfo']: (msg, ws) => {
        this.onUserInfo(msg, ws);
      },
      ['PluginActionConfigurationSchemeResponse']: (msg, ws) => {
        this.onPluginActionConfigurationSchemeResponse(msg, ws);
      },
      ['GetUserInfoList']: (msg, ws) => {
        this.onGetUserInfoList(msg, ws);
      },
      ['GetActionStatus']: (msg, ws) => {
        this.onGetActionStatus(msg, ws);
      },
      ['ActionStatusList']: (msg, ws) => {
        this.onActionStatusList(msg, ws);
      },
      ['ActionStatus']: (msg, ws) => {
        this.onActionStatus(msg, ws);
      },
    };
  }

  #getProto = (type, obj) => {
    switch (type) {
      case 'Envelope':
        return new Envelope(obj)
      case 'PluginHello':
        return new PluginHello(obj)
      case 'Pong':
        return new Pong(obj)
      case 'Settings':
        return new Settings(obj)
      case 'LogEvent':
        return new LogEvent(obj)
      case 'SetData':
        return new SetData(obj)
      case 'GetData':
        return new GetData(obj)
      case 'SettingsRequest':
        return new SettingsRequest(obj)
      case 'WipeData':
        return new WipeData(obj)
      case 'ResponseInfo':
        return new ResponseInfo(obj)
      case 'PluginActionConfigurationSchemeResponse':
        return new PluginActionConfigurationSchemeResponse(obj)
      case 'AnalogControlStatus':
        return new AnalogControlStatus(obj)
      case 'ActionCellStatus':
        return new ActionCellStatus(obj)
      case 'UserInfo':
        return new UserInfo(obj);
      case 'AnalogControlStatusList':
        return new AnalogControlStatusList(obj);
      case 'VisibilityChangedList':
        return new VisibilityChangedList(obj);
      case 'GetActionStatus':
        return new GetActionStatus(obj);
      case 'ActionStatusList':
        return new ActionStatusList(obj);
      case 'ActionStatus':
        return new ActionStatus(obj);
    }
  }

  getType = (obj, type) => {
    return {
      ...obj,
      "@type": `type.googleapis.com/logi.plugin.protocol.${type}`
    }
  };

  _getProtoMessage = (protoName, messageId,  obj= {}) => {
    const protoMessage = this.#getProto(protoName, obj);
    const envelop = this.#getProto('Envelope', {id: messageId});
    envelop.message = protoMessage;
    const envelopObj = JSON.parse(envelop.toJsonString());
    envelopObj.message = this.getType(protoMessage, protoName)

    return JSON.stringify(envelopObj);
  }

  _getProtoMessageWithResponse = (protoName, messageId, obj = {}, err = {}) => {
    const envelop = this.#getProto('Envelope', {id: messageId});
    const envelopObj = JSON.parse(envelop.toJsonString());
    if (protoName !== 'ResponseInfo') {
      const protoMessage = this.#getProto(protoName, obj);
      envelop.message = protoMessage;
      envelopObj.message = this.getType(protoMessage, protoName);
    }
    envelopObj.response = {
      code: (err.code) ? err.code : 1,
      what: (err.what) ? err.what : ''
    }
    return JSON.stringify(envelopObj);
  }

  #defaultResponse = (msg, ws) => {
    const response = this._getProtoMessageWithResponse('ResponseInfo', msg.id, {}, {
      code: -5,
      what: 'Not Implemented'
    });
    try {
      ws.send(response);
    } catch (e) {
      console.log('Something went wrong with ResponseInfo', e);
    }
  }

  #onPing(msg, ws) {
    const response = this._getProtoMessageWithResponse('Pong', msg.id);
    try {
      ws.send(response);
    } catch (e) {
      console.log('Something went wrong with Pong', e);
    }
  }

  onHello() {}

  onTriggerAction(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onPluginActionConfigurationSchemeRequest(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onPluginActionConfigurationResponse(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onRetryAfter(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onReleaseAction(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onVisibilityChanged(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onUpdateAnalogControl(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onSettings(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onSavedData(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onOauthResponse(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onGetAnalogControlValue(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onInitLogin(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onLogout(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onResponseInfo(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onAnalogControlStatus(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onAnalogControlStatusList(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onVisibilityChangedList(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onActionCellStatus(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onUserInfo(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onPluginActionConfigurationSchemeResponse(msg, ws) {
    this.#defaultResponse(msg, ws);
  }

  onGetUserInfoList(msg, ws) {
    this.#defaultResponse(msg, ws);
  }
  onGetActionStatus(msg, ws) {
    this.#defaultResponse(msg, ws);
  }
  onActionStatusList(msg, ws) {
    this.#defaultResponse(msg, ws);
  }
  onActionStatus(msg, ws) {
    this.#defaultResponse(msg, ws);
  }
}

export default Library;